core/koin-core/src/commonMain/kotlin/org/koin/core/Koin.kt

```diff
@@ -324,14 +324,36 @@ class Koin {
      * Load module & create eager instances
      */
     fun loadModules(modules: List<Module>, allowOverride : Boolean = true) {
-        instanceRegistry.loadModules(modules, allowOverride)
-        scopeRegistry.loadScopes(modules)
+        val flattedModules = modules.flatten()
+        instanceRegistry.loadModules(flattedModules, allowOverride)
+        scopeRegistry.loadScopes(flattedModules)
         createEagerInstances()
     }
 
-
     fun unloadModules(modules: List<Module>) {
-        instanceRegistry.unloadModules(modules)
+        val flattedModules = modules.flatten()
+        instanceRegistry.unloadModules(flattedModules)
+    }
+
+    /**
+     * Returns a single list of all [Module] with their [includedModules] in the given list.
+     * Duplicated modules are ignored.
+     */
+    private fun List<Module>.flatten(): List<Module> {
+        val deque = ArrayDeque(elements = this)
+
+        val allModules = mutableListOf<Module>()
+        while (deque.isNotEmpty()) {
+            val module = deque.removeFirstOrNull()
+
+            // We want to ensure duplicated modules aren't visited more than once.
+            if (module != null && !allModules.contains(module)) {
+                allModules += module
+                deque += module.includedModules
+            }
+        }
```

Comment:
Why not a `map` operation here? wouldn't be simpler to map a module and its included modules to a flat list?

Also, what about inner included modules?

Reply:
I am not sure how a `map` would help here. Maybe you can expand a little bit there? 

Note that I'm considering the inner included modules (see below).

Basically, I'm considering each `module` in the list as a possible n-ary tree where each node can include sub-nodes. To iterate over it I'm doing a [Breadth-First Search](https://en.wikipedia.org/wiki/Breadth-first_search): for each node, I iterate over, I will include all `includedModules` to the `DequeArray` that will ensure all `includedModules` are in fact included. Any module that is already included won't be visited again. 

Maybe an image might be a little more helpful to understand what is going on here:

![image](https://user-images.githubusercontent.com/4348197/141983937-081f28e8-0e38-46e9-a471-b4cbf69d4e79.png)

That is the best approach I know to do such operations in an n-ary tree like this one. If you are familiar with another approach, let me know and I will be happy to change it.